src/flag_gems/ops/weightnorm.py

```diff
@@ -0,0 +1,289 @@
+import logging
+import math
+
+import torch
+import triton
+import triton.language as tl
+
+from ..utils import libentry
+
+
+def cfggen_first():
+    block_m = [1, 2, 4, 8, 32]
+    block_n = [512, 1024, 2048]
+    warps = [4, 8, 16]
+    configs = [
+        triton.Config({"BLOCK_ROW_SIZE": m, "BLOCK_COL_SIZE": n}, num_warps=w)
+        for m in block_m
+        for n in block_n
+        for w in warps
+    ]
+    return configs
+
+
+def cfggen_last():
+    block_m = [512, 1024, 2048]
+    block_n = [1, 2, 4, 8, 32]
+    warps = [4, 8, 16]
+    configs = [
+        triton.Config({"BLOCK_ROW_SIZE": m, "BLOCK_COL_SIZE": n}, num_warps=w)
+        for m in block_m
+        for n in block_n
+        for w in warps
+    ]
+    return configs
+
+
+@libentry()
+@triton.autotune(configs=cfggen_last(), key=["M", "N"])
+@triton.jit(do_not_specialize=["eps"])
+def weight_norm_kernel_last(
+    output,
+    norm,
+    v,
+    g,
+    M,
+    N,
+    eps: tl.constexpr,
+    BLOCK_ROW_SIZE: tl.constexpr,
+    BLOCK_COL_SIZE: tl.constexpr,
+):
+    tx = tl.arange(0, BLOCK_COL_SIZE)[:, None]
+    bx = tl.program_id(axis=0) * BLOCK_COL_SIZE
+    col_offset = bx + tx
+    col_mask = col_offset < N
+
+    ty = tl.arange(0, BLOCK_ROW_SIZE)[None, :]
+    v_block = tl.zeros([BLOCK_COL_SIZE, BLOCK_ROW_SIZE], dtype=tl.float32)
+    for base in range(0, M, BLOCK_ROW_SIZE):
+        row_offset = base + ty
+        mask = row_offset < M and col_mask
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask)
+        v_block += v_value * v_value
+
+    normalized = tl.sqrt(tl.sum(v_block, axis=1) + eps)
+    tl.store(norm + col_offset, normalized[:, None], mask=col_mask)
+    g_value = tl.load(g + col_offset, mask=col_mask)
+
+    for base in range(0, M, BLOCK_ROW_SIZE):
+        row_offset = base + ty
+        mask = row_offset < M and col_mask
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask)
+        v_vec = v_value / normalized[:, None]
+        out = v_vec * g_value
+        tl.store(output + row_offset * N + col_offset, out, mask=mask)
+
+
+@libentry()
+@triton.autotune(configs=cfggen_first(), key=["M", "N"])
+@triton.jit(do_not_specialize=["eps"])
+def weight_norm_kernel_first(
+    output,
+    norm,
+    v,
+    g,
+    M,
+    N,
+    eps: tl.constexpr,
+    BLOCK_ROW_SIZE: tl.constexpr,
+    BLOCK_COL_SIZE: tl.constexpr,
+):
+    ty = tl.arange(0, BLOCK_ROW_SIZE)[:, None]
+    by = tl.program_id(axis=0) * BLOCK_ROW_SIZE
+    row_offset = by + ty
+    row_mask = row_offset < M
+
+    tx = tl.arange(0, BLOCK_COL_SIZE)[None, :]
+    v_block = tl.zeros([BLOCK_ROW_SIZE, BLOCK_COL_SIZE], dtype=tl.float32)
+    for base in range(0, N, BLOCK_COL_SIZE):
+        col_offset = base + tx
+        mask = col_offset < N and row_mask
+        v_value = tl.load(v + row_offset * N + col_offset, mask=mask)
+        v_block += v_value * v_value
+
+    normalized = tl.sqrt(tl.sum(v_block, axis=1) + eps)
```

Comment:
I think we should be reducing on the first dimension, ie., axis=0

Reply:
Isn't v_block row-major? summing along dim=1 is more efficient?

Reply:
v_block is stored in row-major order, so I perform the sum along the rows regardless of whether the reduction dimension is the first or last (xy index will be permuted for last). The test encountered an error because REDUCTION_SHAPES = (200, 40999, 3) and dim = 1 is not supported for weight normalization; this issue has now been resolved.

Reply:
Reducing on dim 1 is only correct provided the inputs are transposed up front. It looks like that's not the case in WeightNorm.forward. Can we further verify that?

Reply:
If reduction occurs in the error dimension, the result will definitely be different compared to the golden reference, but currently, they are consistent. The transpose occurs within the kernel, where threads load the number in the row direction from global, but store it in the column direction of v_block.

Reply:
```
            M = v.shape[0]
            N = math.prod(v.shape[1:])
            grid = lambda META: (triton.cdiv(M, META["BLOCK_ROW_SIZE"]),)
```
Above is the blocking scheme in the code,  where M is the reduction dim size. It's clear the reduction axis is split. I don't see how transpose could be done in the kernel...

Reply:
in the kernel
```
// for reduce dim is first
tx = tl.arange(0, BLOCK_COL_SIZE)[None, :]
// for reduce dim is last
ty = tl.arange(0, BLOCK_ROW_SIZE)[None, :]
```
you can verify with a simple instance, for example reduce shape = (2, 2). if reduce dim is wrong in the kernel, the result will not consistent with golden

Reply:
My bad... I took for granted that the input `dim` is the dimension to be contracted off..